src/ol/expr/gpu.js

```diff
@@ -224,6 +226,18 @@ const compilers = {
     const prefix = context.inFragmentShader ? 'v_prop_' : 'a_prop_';
     return prefix + propName;
   },
+  [Ops.Id]: (context) => {
+    const isExisting = FEATURE_ID_PROPERTY_NAME in context.properties;
+    if (!isExisting) {
```

Comment:
Ideally we would only include feature properties in the `context.properties` object. Instead of adding things like the feature id and geometry type to that object, those could be accessed at `context.featureId` and `context.geometryType`.

I understand if this is more that you want to address in this change. But I wanted to make clear how the context object was intended to be used.